### DISCUSSION TITLE
GraphQL resolvers are running in okhttp threadpool which is different from root resolvers threadpool

### DISCUSSION THREAD

**Discussion by tara-9** [graphql-java-kickstart/graphql-java-servlet#461] [Q&A]
Hii team,

GraphqlRootResolver are running in a user defined threadpool. But GraphqlResolvers  are running in OkHttp Thread Pool. The Mdc context is removed for http threadpool. So My Questions are

1.  When GraphQL Query Resolver are Running in user defined threadpool why field resolvers are n't running in that?
2. If it is the expected behaviour how can we copy the context to those thread pools?

My User defined Threadpool Looks Like this 
```
@Bean(name = "graphqlAsyncTaskExecutor")
    public Executor newExecutor() {
        var executor = new ThreadPoolTaskExecutor();
        executor.setCorePoolSize(Runtime.getRuntime().availableProcessors());
        executor.setMaxPoolSize(Runtime.getRuntime().availableProcessors());
        executor.setKeepAliveSeconds(0);
        executor.initialize();
        executor.setThreadNamePrefix("hub-orch");
        return new DelegatingSecurityContextAsyncTaskExecutor(executor);
    }
```
MdcTaskDecorator code
```
public class MdcContextTaskDecorator implements AsyncTaskDecorator, TaskDecorator {

    /**
     * Propagate the current thread's MDC context to the target thread.
     */
    @Override
    public Runnable decorate(Runnable runnable) {
        var mdcContext = MDC.getCopyOfContextMap();
        return () -> {
            try {
                MDC.setContextMap(mdcContext);
                runnable.run();
            } finally {
                MDC.clear();
            }
        };
    }

}
```

**Comment by tara-9** [comment]
After Some debugging i have found the reason why it happens. When there is a nested data object inside another data object the nested data object is completed in the thread in which the outer object is completed.

For eg:  `Author` is a data obj and inside author we have another data obj `books`. Now let's say both are in different service and both require http call to get the data and `books` needs an `author id` to be completed.  we call author api to get the data and it runs in okhttp threadpool then the next book api call will happen in okhttp threadpool. we solved the problem by resolving author in a sperate thread and setting the mdc context for that thread.

But one issue we faced we couldn't do it on the main query resolver thread. In the above e.g let's say another `User` obj exits which wraps `authors` and it is the `root query resolver`. Then what happens is UserQueryResolver completes in `graphql-exex-1` but field resolver will complete in `graphql-exec-2` or `graphql-exec-1` which is random. It is fine but if there is better approach in which we can always completes the field resolvers in the root query resolver thread it will be great.

Thanks.

**Reply by oliemansm** [reply]
Thanks for the clear investigation and analysis.
